File: controllers/books.rb
get "/admin/books" do
  @user = session[:current_user]
  if !@user || !@user.is_admin
    redirect "/403"
  end

  @books = Book.all
  @book_authors = BookAuthor.all
  @book_genres = BookGenre.all
  @authors = Author.all
  @genres = Genre.all

  @active_rentals = Rental.where(is_active: true).order('created_at DESC')
  @old_rentals = Rental.where(is_active: false).order('updated_at DESC')
  @users = User.all
  @book_author = BookAuthor.all
  @book_genre = BookGenre.all

  gon.authors = @authors
  gon.books = @books
  gon.genres = @genres
  gon.book_author = @book_author
  gon.book_genre = @book_genre
  gon.active_rentals = @active_rentals
  gon.old_rentals = @old_rentals 
  gon.users = @users
  haml :'admin/adminBooks'
  #end
end

post "/admin/books" do
  @user = session[:current_user]
  if !@user || !@user.is_admin
    redirect "/403"
  end
  book = Book.create(isbn: params[:isbn].to_i,
    title: params[:title],
    summary: params[:summary])
  params[:authors].each do |a|
    book_author = BookAuthor.create(
      book_id: book.id,
      author_id: a)
  end
  params[:genres].each do |g|
    book_genre = BookGenre.create(
      book_id: book.id,
      genre_id: g)
  end
  redirect "/admin/books"
end

put "/admin/books" do
  @user = session[:current_user]
  if !@user || !@user.is_admin
    redirect "/403"
  end
  book = Book.find(params[:book_id])
  book.update(isbn: params[:isbn].to_i,
    title: params[:title],
    summary: params[:summary])

  book_authors = BookAuthor.where(book_id: book.id)
  book_authors.each do |ba|
    ba.delete()
  end

  book_genres = BookGenre.where(book_id: book.id)
  book_genres.each do |bg|
    bg.delete()
  end

params[:authors].each do |a|
  book_author = BookAuthor.create(
      book_id: book.id,
      author_id: a)
  end
params[:genres].each do |g|
  book_genre = BookGenre.create(
      book_id: book.id,
      genre_id: g)
end

redirect "/admin/books"

end

delete "/admin/books" do
  @user = session[:current_user]
  if !@user || !@user.is_admin
    redirect "/403"
  end
  book = Book.find(params[:book_id])
  if !book
    redirect "/403"
  else
    book.destroy()
    redirect "/admin/books"
  end
end

File: public/js/books.js
function statsBookModal(book_id){
  var active_rentals = $.grep(gon.active_rentals, function(e){ return e.book_id == book_id; });
  var old_rentals = $.grep(gon.old_rentals, function(e){ return e.book_id == book_id; });
  var book_author = $.grep(gon.book_author, function(e){ return e.book_id == book_id; });
  var book_genre = $.grep(gon.book_genre, function(e){ return e.book_id == book_id; });

  $('#activeList').empty();
  $('#historyList').empty();
  $('#authorList').empty();
  $('#genreList').empty();

  $(active_rentals).each(function() {
    var user_id = this.user_id;
    var user = $.grep(gon.users, function(e){ return e.id == user_id; });
    $("#activeList").append('<li>['+user[0].library_card+'] '+ (new Date(this.created_at)).toDateString() +'</li>');
  });
  $(old_rentals).each(function() {
    var user_id = this.user_id;
    var user = $.grep(gon.users, function(e){ return e.id == user_id; });
    $("#historyList").append('<li>['+user[0].library_card+'] '+ (new Date(this.created_at)).toDateString() +'</li>');
  });

  $(book_author).each(function() {
    var author_id = this.author_id;
    var author = $.grep(gon.authors, function(e){ return e.id == author_id; });
    $("#authorList").append('<li>'+ author[0].author_name +'</li>');
  });
  $(book_genre).each(function() {
    var genre_id = this.genre_id;
    var genre = $.grep(gon.genres, function(e){ return e.id == genre_id; });
    $("#genreList").append('<li>'+ genre[0].genre_name +'</li>');
  });

}

function editBooksModal(book_id){
  var book = $.grep(gon.books, function(e){ return e.id == book_id; });
  var book_author = $.grep(gon.book_author, function(e){ return e.book_id == book_id; });
  var book_genre = $.grep(gon.book_genre, function(e){ return e.book_id == book_id; });

  $('#editBookID').val(book[0].id);
  $('#editBookISBN').val(book[0].isbn);
  $('#editBookTitle').val(book[0].title);
  $('#editBookSummary').val(book[0].summary);

  console.log( $('#editBookID').val() );
  $('#editBookGenre').select2("val", ""); 
  $('#editBookAuthor').select2("val", ""); 

  var authors = [];
  $(book_author).each(function() {
    var author_id = this.author_id;
    var author = $.grep(gon.authors, function(e){ return e.id == author_id; });
    var obj = new Object();
    obj.id = author[0].id;
    obj.text  = author[0].author_name;
    authors.push( obj );
  });
  
  $("#editBookAuthor").select2("data", authors );
  var genres = [];
  $(book_genre).each(function() {
    var genre_id = this.genre_id;
    var genre = $.grep(gon.genres, function(e){ return e.id == genre_id; });
    var obj = new Object();
    obj.id = genre[0].id;
    obj.text  = genre[0].genre_name;
    genres.push( obj );
  });
  $("#editBookGenre").select2("data", genres );

}

function deleteBookModal(book_id){
  $('#deleteBookID').val(book_id);
}

$(document).ready(function() {
  $('#bookTable').DataTable({"autoWidth": false});

  $("#addBookAuthor").select2({
    placeholder: "Authors",
    allowClear: true
  });

  $("#addBookGenre").select2({
    placeholder: "Genres",
    allowClear: true
  });

  $("#editBookGenre").select2({
    placeholder: "Genres",
    allowClear: true
  });

  $("#editBookAuthor").select2({
    placeholder: "Authors",
    allowClear: true
  });
} );
